walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -24,64 +24,121 @@
+                if (toAdd.duration() >= request.getDuration()) {
+                    timeRanges.add(toAdd);
+                }
+                current = timeRange;
             }
 
-            if(current.overlaps(attendeeRelevantTimeRanges.get(i))){
-                current = attendeeRelevantTimeRanges.get(i);
-                continue;                
-            }    
+            // Add a TimeRange with a range from the end of the latest event to the end of the day if duration is valid
+            toAdd = TimeRange.fromStartEnd(current.end(), TimeRange.END_OF_DAY, true);
+            if (toAdd.duration() >= request.getDuration()) {
+                timeRanges.add(toAdd);
+            }
 
-            toAdd = TimeRange.fromStartEnd(current.end(), attendeeRelevantTimeRanges.get(i).start(), false);
-            if(toAdd.duration() >= request.getDuration()){
+        } else {
+            current = attendeeRelevantTimeRanges.get(0);
+
+            toAdd = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, current.start(), false);
+            if (toAdd.duration() >= request.getDuration()) {
                 timeRanges.add(toAdd);
             }
-            current = attendeeRelevantTimeRanges.get(i);
-        }
+            
+            for (TimeRange timeRange: attendeeRelevantTimeRanges) {
+                if (current.contains(timeRange)) {
+                    continue;
+                }
 
-        // Add a TimeRange with a range from the end of the latest event to the end of the day if duration is valid
-        toAdd = TimeRange.fromStartEnd(current.end(), TimeRange.END_OF_DAY, true);
-        if(toAdd.duration() >= request.getDuration()){
-            timeRanges.add(toAdd);
-        }
+                if (current.overlaps(timeRange)) {
+                    current = timeRange;
+                    continue;
+                }
+
+                toAdd = TimeRange.fromStartEnd(current.end(), timeRange.start(), false);
+                if (toAdd.duration() >= request.getDuration()) {
+                    timeRanges.add(toAdd);
+                }
+                current = timeRange;
+            }
 
+            toAdd = TimeRange.fromStartEnd(current.end(), TimeRange.END_OF_DAY, true);
+            if (toAdd.duration() >= request.getDuration()) {
+                timeRanges.add(toAdd);
+            }
+
+            List<TimeRange> timeRangesWithNoOptionalInterference = new ArrayList<TimeRange>();
+
+            // Goes through current timeRanges and removes meetings where optionalAttendees can't attend
+            for (int i = timeRanges.size() - 1; i >= 0; i--) {
+                for (TimeRange timeRange: optionalAttendeeRelevantTimeRanges) {
+                    System.out.println(timeRanges.get(i).start() + " " + timeRanges.get(i).end() + " | " + timeRange.start() + " " + timeRange.end());
+                    if (timeRange.overlaps(timeRanges.get(i)) || timeRange.contains(timeRanges.get(i))) {
```

Comment:
Minor improvement, but overlaps and contains run multiple checks (start+end checks for each of them), totaling to 4 checks, but the combined || statement is the same as just saying timerange.contains(timeranges.get(i).start()) || timeranges.contains(timeranges.get(i).end()). The issue is less about the true efficiency, as the fact that the simplist implementation is usually the easiest to understand.